### PR TITLE
FEAT #412: Roles de Usuários no Grupo

### LINKED ISSUE #412: [FEAT] Roles de Usuários no Grupo
# -- Feature --

## Descrição

Funcionalidade de criar papeis de usuários com permissões personalizadas e atribuir papeis para qualquer participante no Grupo.

### PR DESCRIPTION
**Pull request by julio-ufpb**
## Related Issue

Closes #412 #418 
## Proposed Changes

  -
  -
  -

### CODE REVIEW

File: src/pages/Group/GroupConfig/Roles/RolesPage.tsx [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+    const [manageRolesMode, setManageRolesMode] = useState(false);
+
+    const [editMode, setEditMode] = useState(true);
+
+    const [showRolesForm, setShowRolesForm] = useState(false);
+    const [rolesEdit, setRolesEdit] = useState(null as Roles | null);
+
+    const [rows, setRows] = useState(null as Roles[] | null);
+    const [columns, setColumns] = useState(null as string[] | null);
```
Comment (NiiMiyo) [inline src/pages/Group/GroupConfig/Roles/RolesPage.tsx:45]: Alguns desses hooks não são usados em nenhum momento. Vale a pena dar uma analisada nisso pra remover ou usar eles onde necessário.

**Review by NiiMiyo — CHANGES_REQUESTED** [review]

File: src/pages/singin/SinginForm.tsx [hunk 2]
```diff
@@ -38,11 +40,11 @@ export default function SinginForm() {
...
-  const SIGNUP_ENABLED = organizationEnv.signup_enabled ?? true;
-  const ENABLE_GOOGLE_LOGIN = organizationEnv.login_google_enabled ?? (import.meta.env.VITE_ENABLE_GOOGLE_LOGIN === "true" || import.meta.env.VITE_ENABLE_GOOGLE_LOGIN === "1");
-  const ENABLE_RECAPTCHA = organizationEnv.recaptcha_enabled ?? (import.meta.env.VITE_ENABLE_RECAPTCHA === "true" || import.meta.env.VITE_ENABLE_RECAPTCHA === "1");
-  const RECAPTCHA_SITE_KEY = organizationEnv.recaptcha_site_key ?? import.meta.env.VITE_RECAPTCHA_SITE_KEY;
+  let organizationEnv = (((auth.organization??{} as any).groupSettings??{} as any).environment??{} as any);
+  let SIGNUP_ENABLED = organizationEnv.signup_enabled ?? true;
+  let ENABLE_GOOGLE_LOGIN = organizationEnv.login_google_enabled ?? (import.meta.env.VITE_ENABLE_GOOGLE_LOGIN === "true" || import.meta.env.VITE_ENABLE_GOOGLE_LOGIN === "1");
+  let ENABLE_RECAPTCHA = organizationEnv.recaptcha_enabled ?? (import.meta.env.VITE_ENABLE_RECAPTCHA === "true" || import.meta.env.VITE_ENABLE_RECAPTCHA === "1");
+  let RECAPTCHA_SITE_KEY = organizationEnv.recaptcha_site_key ?? import.meta.env.VITE_RECAPTCHA_SITE_KEY;
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/singin/SinginForm.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/360b90271b6ff947c3f426d2a967b7fa9301b19a

File: src/types/Roles.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+
+export enum Permission {
+    DEFAULT = 4,
+    NONE = 0,
+    DISABLED = 1,
+    READ = 2,
+    READ_WRITE = 3,
+    READ_WRITE_DELETE = 4,
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/types/Roles.ts:7]: https://github.com/universi-me/universi-front/pull/449/commits/e65a435f09ee66bf73cb1abe42d7e78e7a73d145

File: src/types/Roles.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+
+
+export type FeatureTypes = "FEED" | "CONTENT" | "GROUP" | "PEOPLE" | "COMPETENCE";
+
+export const FeatureTypesToLabel = {
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/types/Roles.ts]: https://github.com/universi-me/universi-front/pull/449/commits/6e0f303fd244a6fe1af54687a21f31137cfdff8b

File: src/pages/Group/GroupConfig/Roles/RolesLoader.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        return FAILED_TO_LOAD;
+    }
+
+    return {
+        roles: roles.body?.roles,
+        participants: participants.body?.participants,
+    }
+}
+
+export async function RolesLoader(args: LoaderFunctionArgs) {
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/GroupConfig/Roles/RolesLoader.ts]: https://github.com/universi-me/universi-front/pull/449/commits/522426386f6a33a72b033b751e3efd904190caf8

File: src/pages/Group/GroupConfig/Roles/RolesLoader.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    const auth = useContext(AuthContext);
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/GroupConfig/Roles/RolesLoader.ts]: https://github.com/universi-me/universi-front/pull/449/commits/522426386f6a33a72b033b751e3efd904190caf8

File: src/pages/Group/GroupConfig/Roles/RolesLoader.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    const organization = auth.organization;
+
+    if (auth.profile === null || organization === null) {
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/GroupConfig/Roles/RolesLoader.ts]: https://github.com/universi-me/universi-front/pull/449/commits/522426386f6a33a72b033b751e3efd904190caf8

File: src/pages/Group/loader.ts [hunk 8]
```diff
@@ -35,12 +38,12 @@ export async function fetchGroupPageData(props: {groupPath: string | undefined})
...
-        UniversimeApi.Group.subgroups({groupId: group.id}),
-        UniversimeApi.Group.participants({groupId: group.id}),
-        UniversimeApi.Group.folders({groupId: group.id}),
+        canI('GROUP',   Permission.READ, group, profile) == true ? UniversimeApi.Group.subgroups({groupId: group.id})    : {} as any,
+        canI('PEOPLE',  Permission.READ, group, profile) == true ? UniversimeApi.Group.participants({groupId: group.id}) : {} as any,
+        canI('CONTENT', Permission.READ, group, profile) == true ? UniversimeApi.Group.folders({groupId: group.id})      : {} as any,
         UniversimeApi.Profile.groups({profileId: profile.id}),
         UniversimeApi.Profile.links({profileId: profile.id}),
-        UniversimeApi.Feed.getGroupPosts({groupId: group.id}),
+        canI('FEED',    Permission.READ, group, profile) == true ? UniversimeApi.Feed.getGroupPosts({groupId: group.id}) : {} as any,
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/loader.ts]: https://github.com/universi-me/universi-front/pull/449/commits/ae701122d81391bbc5f7a9aa9eaebe94c24977bb

File: src/pages/Group/GroupTabs/GroupTabs.tsx [hunk 9]
```diff
@@ -52,19 +55,21 @@ export  function GroupTabs(props: GroupTabsProps){
...
             TABS.map(t => {
                 const isCurrentTab = t.value === props.currentTab;
 
-                return (
-                    <button className={`group-tab-button`} value={t.value} key={t.value} onClick={() => { window.location.hash = t.value; props.changeTab(t.value); }} data-current-tab={isCurrentTab ? "" : undefined}>
-                        {t.name}
-                    </button>
-                );
+                return (<>
+                    { canI(t!.featureType!, Permission.READ, context?.group) &&
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/GroupTabs/GroupTabs.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/f72b37408826615c5215d5507d210ac88fec9260

File: src/pages/Group/GroupTabs/GroupFeed/GroupFeed.tsx [hunk 10]
```diff
@@ -40,7 +42,7 @@ export function GroupFeed(){
                 <div className="go-right">
                     <Filter setter={setFilterPosts} placeholderMessage={`Buscar publicação em ${groupContext.group.name}`}/>
                     {
-                        canCreatePost()
+                        canCreatePost() || canI("FEED", Permission.READ_WRITE, groupContext.group)
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/GroupTabs/GroupFeed/GroupFeed.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/845fd5208117165577daee352eb68e6e9218c996

File: src/contexts/Auth/AuthProvider.tsx [hunk 11]
```diff
@@ -61,18 +70,24 @@ export const AuthProvider = ({ children }: { children: ReactNode }) => {
...
+        await updateRoles();
         setFinishedLogin(false);
         const profile = await getLoggedProfile();
         setProfile(profile);
 
+        const organization = await updateOrganization();
+
         if (profile !== null) await Promise.all([
-            updateOrganization(),
+            organization,
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/contexts/Auth/AuthProvider.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/9422b37e3352261a94575d0a5c8be661eddd0f74

File: src/pages/Group/GroupTabs/GroupFeed/GroupFeedPost.tsx [hunk 12]
```diff
@@ -73,7 +76,8 @@ export function GroupFeedPost({ post }: GroupFeedPostProps) {
             className: "delete",
             onSelect: handleDeletePost,
             hidden(data) {
-                return !groupContext.group.canEdit && !data.author.user.ownerOfSession;
+                let canDelete = canI("FEED", Permission.READ_WRITE_DELETE, groupContext.group) || (groupContext.group.canEdit || data.author.user.ownerOfSession);
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/pages/Group/GroupTabs/GroupFeed/GroupFeedPost.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/01014163e6f50a4b722b82d3b60b7ca4a9dd9af9

File: src/contexts/Auth/AuthContext.tsx [hunk 13]
```diff
@@ -10,6 +10,7 @@ export type AuthContextType = {
     profileLinks: Link[];
     profileGroups: Group[];
     organization: Group | null;
+    roles: any;
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/contexts/Auth/AuthContext.tsx]: https://github.com/universi-me/universi-front/blob/669e76faf77f507e93bdb2b8ece678852e4c5f6b/src/types/Roles.tshttps://github.com/universi-me/universi-front/pull/449/commits/669e76faf77f507e93bdb2b8ece678852e4c5f6b

File: src/types/Roles.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    "COMPETENCE":   "Competência"
+};
+
+export type Roles = {
+    id: string;
+    name: string;
+    description : string;
+    isDefault: boolean;
+    rolesFeatures: RolesFeature[];
+    features?: RolesFeature[];
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/types/Roles.ts]: https://github.com/universi-me/universi-front/pull/449/commits/669e76faf77f507e93bdb2b8ece678852e4c5f6b

File: src/contexts/Auth/AuthContext.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/contexts/Auth/AuthContext.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/669e76faf77f507e93bdb2b8ece678852e4c5f6b

File: src/utils/roles/rolesUtils.ts [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+  let roles = getRolesFromLocalStorage();
+  if(roles ===  null || !roles) {
+    // fetch roles from API
+    return UniversimeApi.Roles.listRoles().then((data : any) => {
+      if(data.success && data.body.roles) {
+        saveRolesLocalStorage(data.body.roles);
+      }
+      return canI(featureType, permission, group, profile);
+    });
+  }
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/utils/roles/rolesUtils.ts]: https://github.com/universi-me/universi-front/pull/449/commits/669e76faf77f507e93bdb2b8ece678852e4c5f6b

File: src/utils/roles/rolesUtils.ts [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * Check if the user has permission to do something
+ * @param featureType The feature type
+ * @param permission The permission, if not provided, will return number of permission of user
+ * @param profile The profile
+ * @param group The group
+ * @returns The permission or false if the user does not have permission
+ */
+export function canI(featureType: FeatureTypes, permission?: Permission, group?: Group, profile?: Profile): boolean | number  {
+  let returnValueAsBoolean = (permission != null || permission != undefined);
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/utils/roles/rolesUtils.ts]: https://github.com/universi-me/universi-front/pull/449/commits/669e76faf77f507e93bdb2b8ece678852e4c5f6b

File: src/services/UniversimeApi/Roles.ts [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+export async function assigned(body:  RolesAssigned_RequestDTO) {
+    return (await api.post<RolesRemove_ResponseDTO>("/roles/assigned", {
+        groupId:                 body.groupId,
+        profileId:               body.profileId
+    })).data;
+}
+
+export function listRoles() : any {
```
[the remark this answers is not in the document]
Comment (NiiMiyo) [reply src/services/UniversimeApi/Roles.ts]: https://github.com/universi-me/universi-front/commit/ab0765b528adab9bf96959dac0d3dc15b17c0db5

File: src/pages/Group/GroupConfig/Roles/RolesPage.tsx (hunk 1 above)
Comment (NiiMiyo) [reply src/pages/Group/GroupConfig/Roles/RolesPage.tsx]: https://github.com/universi-me/universi-front/pull/449/commits/669e76faf77f507e93bdb2b8ece678852e4c5f6b

**Review by NiiMiyo — APPROVED** [review]

**Review by 710lucas — APPROVED** [review]

**Pull request merged**